entropylab/api/param_store.py

```diff
@@ -0,0 +1,322 @@
+        return self._params.__iter__()
+
+    def __len__(self) -> int:
+        return self._params.__len__()
+
+    def __setattr__(self, name, value):
+        if name.startswith("_"):
+            super().__setattr__(name, value)
+        else:
+            self._params[name] = value
+            self._is_dirty = True
+
+    def __getattr__(self, name):
+        if name.startswith("_"):
+            return super().__getattribute__(name)
+        else:
+            return self._params[name]
+
+    def __delattr__(self, name):
+        if name.startswith("_"):
+            super().__delattr__(name)
+        else:
+            del self._params[name]
+            self._is_dirty = True
+
+    """ Items """
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        self._params[key] = value
+        self._is_dirty = True
+
+    def __getitem__(self, key: str) -> Any:
+        return self._params[key]
+
+    def __delitem__(self, *args, **kwargs):
+        self._params.__delitem__(*args, **kwargs)
+        self._is_dirty = True
+
+    def __contains__(self, key: str) -> bool:
+        return key in self._params
+
+    def to_dict(self) -> Dict:
+        return dict(self._params)
+
+    def get(self, key: str, commit_id: Optional[str] = None):
+        """
+            returns the value of a param by key
+
+        :param key: the key identifying the param
+        :param commit_id: an optional commit_id. if provided, the value will be
+        returned from the specified commit
+        """
+        if commit_id is None:
+            return self._params.get(key)
+        else:
+            commit = self._get_commit(commit_id)
+            return commit["params"][key]
+
+    """ Commits """
+
+    def commit(self, label: Optional[str] = None) -> str:
+        if not self._is_dirty:
+            return self._base_commit_id
+        metadata = self._generate_metadata(label)
+        if self._is_in_memory_mode:
+            params = self._deep_copy(self._params)
+        else:
+            params = self._params
+        doc_id = self._db.insert(dict(metadata=metadata.__dict__, params=params))
+        self._base_commit_id = metadata.id
+        self._base_doc_id = doc_id
+        self._is_dirty = False
+        return metadata.id
+
+    def checkout(
+        self,
+        commit_id: Optional[str] = None,
+        commit_num: Optional[int] = None,
+        move_by: Optional[int] = None,
+    ) -> None:
+        commit = self._get_commit(commit_id, commit_num, move_by)
+        self._params = commit["params"]
+        self._base_commit_id = commit_id
+        self._base_doc_id = commit.doc_id
+        self._is_dirty = False
+
+    def list_commits(self, label: Optional[str] = None) -> List[Metadata]:
+        """
+            returns a list of commits
+
+        :param label: an optional label, if given then only commits that match
+        it will be returned
+        """
+        documents = self._db.search(
```

Comment:
do we want to skip if the label is None?

Reply:
I don't think so... If no `label` is provided then the user gets a list of all commits in the store. Is there a problem with that?

Reply:
ok, probably I didn't understand the api